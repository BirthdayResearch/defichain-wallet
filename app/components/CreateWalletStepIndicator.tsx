import React from 'react'
import { StyleProp, ViewStyle } from 'react-native'
import { Text, View } from '.'
import { tailwind } from '../tailwind'
import { translate } from '../translations'

interface StepIndicatorProps {
  current: number
  total?: number
  steps?: string[]
  style?: StyleProp<ViewStyle>
}

export const CREATE_STEPS = [
  translate('components/CreateWalletIndicator', 'recovery'),
  translate('components/CreateWalletIndicator', 'verify'),
  translate('components/CreateWalletIndicator', 'passcode')
]

export const RESTORE_STEPS = [
  translate('components/CreateWalletIndicator', 'restore'),
  translate('components/CreateWalletIndicator', 'passcode')
]

/**
 * @param props
 * @param {number} props.total throw error if not fulfill (3 < total < 6), optional when props.steps provided
 * @param {number} props.current throw error if not fulfill (0 < current <= total)
 * @param {string[]} props.steps description displayed for each step
 * @returns {JSX.Element}
 */
export function CreateWalletStepIndicator (props: StepIndicatorProps): JSX.Element {
  const { current, total, style: containerViewStyle, steps = [] } = props
  if (total === undefined && steps.length === 0) {
    throw Error('Invalid prop for CreateWalletStepIndicator')
  }

  const totalStep = total ?? steps.length
  if (totalStep > 5 || current <= 0 || current > totalStep) {
    throw Error('Invalid prop for CreateWalletStepIndicator')
  }

  function following (): JSX.Element[] {
    const arr: JSX.Element[] = []
    for (let i = 1; i < totalStep; i++) {
      const iconStyle = current >= i + 1 ? 'bg-primary' : 'bg-gray-100'
      arr.push(
        <View
          key={i * 2}
          style={tailwind(`h-1 flex-grow mt-3.5 ${iconStyle}`)}
        />)
      arr.push(<StepNode key={i * 2 + 1} step={i + 1} current={current} />)
    }
    return arr
  }

  function descriptions (): JSX.Element[] {
    const arr: JSX.Element[] = []
    for (let i = 0; i < steps.length; i++) {
      let textStyle = ''
      if (current === i + 1) {
        textStyle = 'text-primary'
      } else {
        textStyle = 'text-gray-500'
      }
      arr.push(
        <View key={i}>
          <Text style={tailwind(`text-center text-sm font-medium ${textStyle}`)}>{steps[i]}</Text>
        </View>
      )
    }
    return arr
  }

  return (
    <View style={[tailwind('flex-col justify-center items-center w-full'), containerViewStyle]}>
      <View style={[tailwind('flex-row justify-center w-9/12')]}>
        <StepNode step={1} current={current} />
        {following()}
      </View>
      <View style={[tailwind('flex-row justify-between w-10/12 mt-1 font-medium')]}>
        {descriptions()}
      </View>
    </View>
  )
}

function StepNode (props: { step: number, current: number }): JSX.Element {
  let stepperStyle
  let textStyle
  if (props.current === props.step) {
    stepperStyle = 'bg-primary bg-opacity-10 border border-primary'
    textStyle = 'text-primary'
  } else if (props.current > props.step) {
    stepperStyle = 'bg-primary border border-primary'
    textStyle = 'text-white'
  } else {
    stepperStyle = 'bg-transparent border border-gray-200'
    textStyle = 'text-gray-500'
  }
  return (
    <View style={tailwind('flex-col')}>
      <View
        style={tailwind(`h-8 w-8 rounded-2xl justify-center items-center ${stepperStyle}`)}
      >
        <Text style={tailwind(`${textStyle} font-medium`)}>{props.step}</Text>
      </View>
    </View>
  )
}
