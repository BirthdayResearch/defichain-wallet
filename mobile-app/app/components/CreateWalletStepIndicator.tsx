import { StyleProp, ViewStyle } from 'react-native'
import { Text, View } from '.'
import { useThemeContext } from '@shared-contexts/ThemeProvider'
import { tailwind } from '@tailwind'
import { translate } from '@translations'
import { ThemedText } from './themed'

interface StepIndicatorProps {
  current: number
  total?: number
  steps?: string[]
  style?: StyleProp<ViewStyle>
}

export const CREATE_STEPS = [
  'Recovery',
  'Verify',
  'Secure'
]

export const RESTORE_STEPS = [
  'Restore',
  'Secure'
]

/**
 * @param props
 * @param {number} props.total throw error if not fulfill (3 < total < 6), optional when props.steps provided
 * @param {number} props.current throw error if not fulfill (0 < current <= total)
 * @param {string[]} props.steps description displayed for each step
 * @returns {JSX.Element}
 */
export function CreateWalletStepIndicator (props: StepIndicatorProps): JSX.Element {
  const { isLight } = useThemeContext()
  const {
    current,
    total,
    style: containerViewStyle,
    steps = []
  } = props
  if (total === undefined && steps.length === 0) {
    throw Error('Invalid prop for CreateWalletStepIndicator')
  }

  const totalStep = total ?? steps.length
  if (totalStep > 5 || current <= 0 || current > totalStep) {
    throw Error('Invalid prop for CreateWalletStepIndicator')
  }

  function following (): JSX.Element[] {
    const arr: JSX.Element[] = []
    for (let i = 1; i < totalStep; i++) {
      const iconStyle = isLight ? (current >= i + 1 ? 'bg-primary-500' : 'bg-gray-100') : (current >= i + 1 ? 'bg-darkprimary-400' : 'bg-gray-600')
      arr.push(
        <View
          key={i * 2}
          style={tailwind(`h-1 flex-grow mt-3.5 ${iconStyle}`)}
        />)
      arr.push(
        <StepNode
          content={steps[i]}
          current={current}
          isLight={isLight}
          key={i * 2 + 1}
          step={i + 1}
        />)
    }
    return arr
  }

  return (
    <View style={[tailwind('flex-col justify-center items-center w-full'), containerViewStyle]}>
      <View style={tailwind('flex-row justify-center w-9/12 h-14')}>
        <StepNode
          content={steps[0]}
          current={current}
          isLight={isLight}
          step={1}
        />

        {following()}
      </View>
    </View>
  )
}

function getStepNodeStyle (isLight: boolean, current: number, step: number): { stepperStyle: string, textStyle: string } {
  let stepperStyle
  let textStyle
  if (current === step) {
    stepperStyle = isLight ? 'bg-primary-500 bg-opacity-10 border border-primary-500' : 'bg-darkprimary-300 border border-darkprimary-600'
    textStyle = isLight ? 'text-primary-500' : 'text-darkprimary-700'
  } else if (current > step) {
    stepperStyle = isLight ? 'bg-primary-500 border border-primary-500' : 'bg-darkprimary-500 border border-darkprimary-600'
    textStyle = 'text-white'
  } else {
    stepperStyle = isLight ? 'bg-transparent border border-gray-200' : 'bg-gray-700 border border-gray-200'
    textStyle = isLight ? 'text-gray-500' : 'text-gray-400'
  }
  return {
    stepperStyle,
    textStyle
  }
}

function StepNode (props: { step: number, current: number, content: string, isLight: boolean }): JSX.Element {
  const {
    stepperStyle,
    textStyle
  } = getStepNodeStyle(props.isLight, props.current, props.step)
  return (
    <View style={tailwind('flex-col')}>
      <View
        style={tailwind(`h-8 w-8 rounded-2xl justify-center items-center relative ${stepperStyle}`)}
      >
        <Text style={tailwind(`${textStyle} font-medium absolute`)}>
          {props.step}
        </Text>

        <Description
          content={props.content}
          current={props.current}
          step={props.step}
        />
      </View>
    </View>
  )
}

function Description (props: { step: number, current: number, content: string }): JSX.Element {
  return (
    <ThemedText
      dark={tailwind(props.current === props.step ? 'text-darkprimary-400' : 'text-gray-400')}
      light={tailwind(props.current === props.step ? 'text-primary-500' : 'text-gray-500')}
      style={tailwind('text-center text-sm font-medium top-9 absolute w-20')}
    >
      {translate('components/CreateWalletIndicator', props.content)}
    </ThemedText>
  )
}
