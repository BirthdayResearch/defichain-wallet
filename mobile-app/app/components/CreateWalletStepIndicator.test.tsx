import { render } from '@testing-library/react-native'
import { CreateWalletStepIndicator } from './CreateWalletStepIndicator'

jest.mock('@shared-contexts/ThemeProvider')

describe('Create wallet step indicator', () => {
  it('should match snapshot', () => {
    const rendered = render(
      <CreateWalletStepIndicator
        current={1}
        steps={['Recovery', 'Verify', 'Secure']}
      />
    )
    expect(rendered.toJSON()).toMatchSnapshot()
  })
})
