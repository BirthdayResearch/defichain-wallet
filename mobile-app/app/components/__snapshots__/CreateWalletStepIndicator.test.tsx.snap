// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create wallet step indicator should match snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "column",
        "justifyContent": "center",
        "width": "100%",
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
        "height": 56,
        "justifyContent": "center",
        "width": "75%",
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "column",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "rgba(255, 0, 175, 0.1)",
            "borderBottomColor": "rgba(255, 0, 175, 1)",
            "borderBottomLeftRadius": 16,
            "borderBottomRightRadius": 16,
            "borderBottomWidth": 1,
            "borderLeftColor": "rgba(255, 0, 175, 1)",
            "borderLeftWidth": 1,
            "borderRightColor": "rgba(255, 0, 175, 1)",
            "borderRightWidth": 1,
            "borderTopColor": "rgba(255, 0, 175, 1)",
            "borderTopLeftRadius": 16,
            "borderTopRightRadius": 16,
            "borderTopWidth": 1,
            "height": 32,
            "justifyContent": "center",
            "position": "relative",
            "width": 32,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontFamily": "RegularFont",
                "fontSize": 16,
                "fontWeight": "400",
                "lineHeight": 24,
              },
              Object {
                "color": "rgba(255, 0, 175, 1)",
                "fontFamily": "MediumFont",
                "fontWeight": "500",
                "position": "absolute",
              },
            ]
          }
        >
          1
        </Text>
        <Text
          style={
            Array [
              Object {
                "fontFamily": "RegularFont",
                "fontSize": 16,
                "fontWeight": "400",
                "lineHeight": 24,
              },
              Array [
                Object {
                  "fontFamily": "MediumFont",
                  "fontSize": 14,
                  "fontWeight": "500",
                  "lineHeight": 20,
                  "position": "absolute",
                  "textAlign": "center",
                  "top": 36,
                  "width": 80,
                },
                Object {
                  "color": "rgba(255, 0, 175, 1)",
                },
              ],
            ]
          }
        >
          Recovery
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "backgroundColor": "rgba(245, 245, 245, 1)",
          "flexGrow": 1,
          "height": 4,
          "marginTop": 14,
        }
      }
    />
    <View
      style={
        Object {
          "flexDirection": "column",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "borderBottomColor": "rgba(229, 229, 229, 1)",
            "borderBottomLeftRadius": 16,
            "borderBottomRightRadius": 16,
            "borderBottomWidth": 1,
            "borderLeftColor": "rgba(229, 229, 229, 1)",
            "borderLeftWidth": 1,
            "borderRightColor": "rgba(229, 229, 229, 1)",
            "borderRightWidth": 1,
            "borderTopColor": "rgba(229, 229, 229, 1)",
            "borderTopLeftRadius": 16,
            "borderTopRightRadius": 16,
            "borderTopWidth": 1,
            "height": 32,
            "justifyContent": "center",
            "position": "relative",
            "width": 32,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontFamily": "RegularFont",
                "fontSize": 16,
                "fontWeight": "400",
                "lineHeight": 24,
              },
              Object {
                "color": "rgba(115, 115, 115, 1)",
                "fontFamily": "MediumFont",
                "fontWeight": "500",
                "position": "absolute",
              },
            ]
          }
        >
          2
        </Text>
        <Text
          style={
            Array [
              Object {
                "fontFamily": "RegularFont",
                "fontSize": 16,
                "fontWeight": "400",
                "lineHeight": 24,
              },
              Array [
                Object {
                  "fontFamily": "MediumFont",
                  "fontSize": 14,
                  "fontWeight": "500",
                  "lineHeight": 20,
                  "position": "absolute",
                  "textAlign": "center",
                  "top": 36,
                  "width": 80,
                },
                Object {
                  "color": "rgba(115, 115, 115, 1)",
                },
              ],
            ]
          }
        >
          Verify
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "backgroundColor": "rgba(245, 245, 245, 1)",
          "flexGrow": 1,
          "height": 4,
          "marginTop": 14,
        }
      }
    />
    <View
      style={
        Object {
          "flexDirection": "column",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "borderBottomColor": "rgba(229, 229, 229, 1)",
            "borderBottomLeftRadius": 16,
            "borderBottomRightRadius": 16,
            "borderBottomWidth": 1,
            "borderLeftColor": "rgba(229, 229, 229, 1)",
            "borderLeftWidth": 1,
            "borderRightColor": "rgba(229, 229, 229, 1)",
            "borderRightWidth": 1,
            "borderTopColor": "rgba(229, 229, 229, 1)",
            "borderTopLeftRadius": 16,
            "borderTopRightRadius": 16,
            "borderTopWidth": 1,
            "height": 32,
            "justifyContent": "center",
            "position": "relative",
            "width": 32,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontFamily": "RegularFont",
                "fontSize": 16,
                "fontWeight": "400",
                "lineHeight": 24,
              },
              Object {
                "color": "rgba(115, 115, 115, 1)",
                "fontFamily": "MediumFont",
                "fontWeight": "500",
                "position": "absolute",
              },
            ]
          }
        >
          3
        </Text>
        <Text
          style={
            Array [
              Object {
                "fontFamily": "RegularFont",
                "fontSize": 16,
                "fontWeight": "400",
                "lineHeight": 24,
              },
              Array [
                Object {
                  "fontFamily": "MediumFont",
                  "fontSize": 14,
                  "fontWeight": "500",
                  "lineHeight": 20,
                  "position": "absolute",
                  "textAlign": "center",
                  "top": 36,
                  "width": 80,
                },
                Object {
                  "color": "rgba(115, 115, 115, 1)",
                },
              ],
            ]
          }
        >
          Secure
        </Text>
      </View>
    </View>
  </View>
</View>
`;
